test/PostageStamp.test.ts

```diff
@@ -42,22 +46,160 @@ describe('PostageStamp', function () {
+      });
+
+      it('should transfer the token', async function () {
+        await this.postageStamp.createBatch(stamper, this.batch.initialPayment, this.batch.depth, this.batch.nonce);
+        expect(await this.token.balanceOf(stamper)).to.equal(0);
+      });
+
+      it('should not create batch if insufficient funds', async function () {
+        expect(
+          this.postageStamp.createBatch(stamper, this.batch.initialPayment + 100, this.batch.depth, this.batch.nonce)
+        ).to.be.revertedWith('ERC20: transfer amount exceeds balance');
+      });
+
+      it('should not allow zero address as owner', async function () {
+        expect(
+          this.postageStamp.createBatch(zeroAddress, this.batch.initialPayment, this.batch.depth, this.batch.nonce)
+        ).to.be.revertedWith('owner cannot be the zero address');
+      });
+
+      it('should not allow duplicate batch', async function () {
+        await this.postageStamp.createBatch(stamper, 0, this.batch.depth, this.batch.nonce);
+        await expect(this.postageStamp.createBatch(stamper, 0, this.batch.depth, this.batch.nonce)).to.be.revertedWith(
+          'batch already exists'
+        );
+      });
+    });
+
+    describe('when topping up a batch', function () {
+      beforeEach(async function () {
+        this.postageStamp = await ethers.getContract('PostageStamp', stamper);
+        this.token = await ethers.getContract('ERC20PresetMinterPauser', deployer);
+
+        this.batch = {
+          nonce: '0x000000000000000000000000000000000000000000000000000000000000abcd',
+          initialPayment: 20,
+          depth: 5,
+        };
+
+        this.batch.id = computeBatchId(stamper, this.batch.nonce);
+        this.topupAmount = 100;
+
+        await this.token.mint(stamper, this.batch.initialPayment + this.topupAmount);
+        (await ethers.getContract('ERC20PresetMinterPauser', stamper)).approve(
+          this.postageStamp.address,
+          this.batch.initialPayment + this.topupAmount
+        );
+        await this.postageStamp.createBatch(stamper, this.batch.initialPayment, this.batch.depth, this.batch.nonce);
+      });
+
+      it('should fire the BatchTopUp event', async function () {
+        await expect(this.postageStamp.topUp(this.batch.id, this.topupAmount))
+          .to.emit(this.postageStamp, 'BatchTopUp')
+          .withArgs(this.batch.id, this.topupAmount);
+      });
+
+      it('should transfer the token', async function () {
+        await this.postageStamp.topUp(this.batch.id, this.topupAmount);
+        expect(await this.token.balanceOf(stamper)).to.equal(0);
+        expect(await this.token.balanceOf(this.postageStamp.address)).to.equal(
+          this.batch.initialPayment + this.topupAmount
+        );
+      });
+
+      it('should not top up non-existing batches', async function () {
+        await expect(
+          this.postageStamp.topUp(computeBatchId(deployer, this.batch.nonce), this.topupAmount)
```

Comment:
Suggestion: extract `computeBatchId(deployer, this.batch.nonce)` to a variable such as `nonExistingBatchId`. That way it's quite clear what is going on regardless of the test setup. Alternatively, we could have a comment just to clarify.

Reply:
fixed